### PR TITLE
cluster manager fixes

### LINKED ISSUE #30031: failed worker startup
even if remote cluster worker succeeds, text is printed which says it fails:

```
julia> using Distributed, ClusterManagers

julia> addprocs_lsf(1)
<<Waiting for dispatch ...>>
<<Starting on h11u01>>
	From failed worker startup:	Job <50204976> is submitted to default queue <interactive>.
1-element Array{Int64,1}:
 2

julia> fetch(@spawnat 2 myid())
2

julia> nprocs()
2
```

i believe all that needs to be changed is removing "failed" from [this line](https://github.com/JuliaLang/julia/blob/453a7ddcb6894cfcd8fc016634857b39416585cc/stdlib/Distributed/src/cluster.jl#L301).

could be something wrong with [this PR](https://github.com/JuliaParallel/ClusterManagers.jl/pull/74) though, which was used in the MWE above.

### PR DESCRIPTION
**Pull request by bjarthur**
fixes https://github.com/JuliaLang/julia/issues/30031

the two commit messages say it all.

i'll happily add a third commit to fix the REPL being blocked during worker startup if someone gives me a hint as to what to change.  this is a problem when the remote worker ends up in the pending queue on a busy cluster.  in julia 0.6 the remote workers were launched in the background.  i've tried making a few changed to the various `@asyncs` that are sprinked through the relevant code, but nothing has worked so far.

### CODE REVIEW

**Comment by bjarthur** [comment]
@amitmurthy could you please review?  thanks.

**Comment by vchuravy** [comment]
This looks okay, but it would be good to have a test

**Comment by ViralBShah** [comment]
@amitmurthy may not be watching.

**Comment by amitmurthy** [comment]
Why don't we just remove "failed" from https://github.com/JuliaLang/julia/blob/a50d5af9cb9a7cce9a5ee3154bba735ffc74c1b1/stdlib/Distributed/src/cluster.jl#L301 and leave everything else as is? It will be good to print all output from the workers prior to reading the host/port info - even the informational stuff.

File: stdlib/Distributed/src/cluster.jl [hunk 1]
```diff
@@ -605,7 +608,12 @@ function launch_additional(np::Integer, cmd::Cmd)
     end
 
     for (i,io) in enumerate(io_objs)
-        (host, port) = read_worker_host_port(io)
+        try
+            (host, port) = read_worker_host_port(io)
+        catch e
+            kill(io)
```
Comment (amitmurthy) [inline stdlib/Distributed/src/cluster.jl]: `kill` needs either a Process object or individual cluster managers need to implement it. I suggest that we throw a `TimedOutException` and handle it specifically in the cluster manager. I think you will need to call a command specific to the particular cluster manager, i.e., specific to LSF, SLURM, etc...

File: stdlib/Distributed/src/managers.jl [hunk 2]
```diff
@@ -394,7 +394,11 @@ function connect(manager::ClusterManager, pid::Int, config::WorkerConfig)
 
     # master connecting to workers
     if config.io !== nothing
-        (bind_addr, port) = read_worker_host_port(config.io)
+        try
+            (bind_addr, port) = read_worker_host_port(config.io)
+        catch e
+            error(e.msg)
```
Comment (JeffBezanson) [inline stdlib/Distributed/src/managers.jl]: What's the reason for this? Looks unnecessary to me, plus could cause a problem if `e` doesn't have a `msg` field.

File: stdlib/Distributed/src/cluster.jl (hunk 1 above)
Comment (bjarthur) [reply stdlib/Distributed/src/cluster.jl]: i've moved the `kill` to  `create_worker()`, which has all the needed info to call the method in ClusterManagers.

File: stdlib/Distributed/src/managers.jl (hunk 2 above)
Comment (bjarthur) [reply stdlib/Distributed/src/managers.jl]: emitting the error msg is now handled by a `finally` block with a `rethrow` in `create_worker()`

**Comment by bjarthur** [comment]
lastly, i've added documentation as how to asynchronously launch workers.

it'd be great if this PR could be reviewed again.  thanks for the previous comments.  much better now.

**Review by vchuravy — APPROVED** [review]
This looks good to me.

File: stdlib/Distributed/src/cluster.jl [hunk 3]
```diff
@@ -354,6 +359,11 @@ the package `ClusterManagers.jl`.
 The number of seconds a newly launched worker waits for connection establishment from the
 master can be specified via variable `JULIA_WORKER_TIMEOUT` in the worker process's
 environment. Relevant only when using TCP/IP as transport.
+
+To launch workers without blocking the REPL, or the containing function
+if launching workers programmatically, execute `addprocs` in its own task.
+The normal output, including any messages, can be subsequently captured by
+fetching the returned Task object: `t = @async addprocs(...); ...;  fetch(t)`.
```
[the remark this answers is not in the document]
Comment (amitmurthy) [reply stdlib/Distributed/src/cluster.jl]: As currently worded, I find it a bit limiting. Is your specific use case "busy" clusters where you want to use workers as and when they come online? Maybe you can highlight that specific use case and give examples on how to use it in that scenario? A few code snippets I could think of concerning your use case...

```
# On busy cluster managers, to call addprocs asynchronously  
t = @async addprocs(...)

# To utilize workers as and when they come online 
if nprocs() > 1   # Ensure at least one new worker is available 
   ....  # perform distributed execution
end

# Check if `addprocs` has completed.
istaskdone(t)

# Check if there were any errors in launching `N` new workers
if nworkers() != N && istaskdone(t)
   fetch(t)   # `istaskdone` check above ensures `fetch` doesn't block                                    
end

# Retrieve newly launched worker ids
if nworkers() == N && istaskdone(t)
   new_pids = fetch(t)   # `istaskdone` check above ensures `fetch` doesn't block                                    
end
```

Overall, I think highlighting the use case of busy clusters where addprocs can take a while for the complete launch, and a need to interactively check its progress (or use resources as and when they come online), with  examples would be better here.

[the remark this answers is not in the document]
Comment (bjarthur) [reply stdlib/Distributed/src/cluster.jl]: your examples for busy clusters added.  anything else?

**Comment by bjarthur** [comment]
thanks!  @amitmurthy could you please review the related https://github.com/JuliaParallel/ClusterManagers.jl/pull/74 too?

**Comment by JeffBezanson** [comment]
We can backport this to 1.0 if it's critical for somebody's cluster environment, but for now we won't plan to.

**Pull request merged**